types/v56/dse.go

```diff
@@ -0,0 +1,124 @@
+	APIVersion string            `json:"apiVersion"`
+}
+
+type DseConfigMetadata struct {
+	Name string `json:"name"`
+}
+
+type DseConfigSpec struct {
+	Artifacts    []DseArtifactMap `json:"artifacts"`
+	Description  string           `json:"description"`
+	DockerConfig *DseDockerConfig `json:"dockerConfig,omitempty"`
+	SolutionType string           `json:"solutionType"`
+}
+
+type DseArtifactMap map[string]interface{}
+
+// DseDockerConfig provides registry auth configuration that is available for "VCD Data Solutions"
+// Data Solution
+type DseDockerConfig struct {
+	Auths DseDockerAuths `json:"auths"`
+}
+
+type DseDockerAuths map[string]DseDockerAuth
+
+type DseDockerAuth struct {
+	Username    string `json:"username"`
+	Password    string `json:"password"`
+	Description string `json:"description"`
+}
+
+// DataSolutionOrgConfig represents RDE Entity structure for Data Solution Org Configuration
+type DataSolutionOrgConfig struct {
+	APIVersion string                 `json:"apiVersion"`
+	Kind       string                 `json:"kind"`
+	Metadata   map[string]interface{} `json:"metadata"`
+	Spec       map[string]interface{} `json:"spec"`
+}
+
+// DataSolutionInstanceTemplate represents a read-only structure for Data Solution Instance
+// Templates in Data Storage Extension (DSE)
+type DataSolutionInstanceTemplate struct {
+	Kind string `json:"kind"`
+	Spec struct {
+		Data struct {
+			CPU              string `json:"cpu"`
+			Name             string `json:"name"`
+			Memory           string `json:"memory"`
+			Storage          string `json:"storage"`
+			Namespace        string `json:"namespace"`
+			PvcPolicy        string `json:"pvcPolicy"`
+			HighAvailability bool   `json:"highAvailability"`
+		} `json:"data"`
+		BuiltIn    bool   `json:"builtIn"`
+		Content    string `json:"content"`
+		Version    string `json:"version"`
+		Featured   bool   `json:"featured"`
+		DataSchema struct {
+			Type string `json:"type"`
+			Defs struct {
+				Quantity struct {
+					Type    string `json:"type"`
+					Pattern string `json:"pattern"`
+				} `json:"quantity"`
+				LimitedQuantity struct {
+					Type    string `json:"type"`
+					Pattern string `json:"pattern"`
+				} `json:"limitedQuantity"`
+			} `json:"$defs"`
+			Schema     string   `json:"$schema"`
+			Required   []string `json:"required"`
+			Properties struct {
+				CPU struct {
+					Ref         string `json:"$ref"`
+					Title       string `json:"title"`
+					Default     string `json:"default"`
+					Description string `json:"description"`
+				} `json:"cpu"`
+				Memory struct {
+					Ref         string `json:"$ref"`
+					Title       string `json:"title"`
+					Default     string `json:"default"`
+					Description string `json:"description"`
+				} `json:"memory"`
+				Storage struct {
+					Ref         string `json:"$ref"`
+					Title       string `json:"title"`
+					Default     string `json:"default"`
+					Description string `json:"description"`
```

Comment:
Aren't we repeating the same structure several times with different names ?

Reply:
Indeed you are right. To be honest I didn't dive deep into this structure as it is read-only, but surely this can be optimised.

Reply:
Extracted at least these 3 elements.